docs: jsx (#3)
pushdev-code/react-tutorial: 2-using-jsx/index.md

# Using JSX

## JSX = JavaScript XML.

> JSX allows us to write HTML in React, making it easier to write and add HTML in React. JSX produces React “elements” and we will render them to the DOM.

Reviewer: ```suggestion
JSX allows us to write HTML in React, making it easier to understand what's happening with the components, just by looking at the JSX part of the component. JSX produces React “elements” and we will render them to the DOM.
```

```js
const message = 'PushDev';
const element = <h1 className='greeting'>Hello, {message}</h1>;

ReactDOM.render(
  element,
  document.getElementById('root')
);
```

> Under the hood it's using `React.createElement()`

Reviewer: ```suggestion
Under the hood it is using `React.createElement()`
```

```js
const message = 'PushDev';
const element = React.createElement('h1', {className: 'greeting'}, `Hello, ${message}`);

ReactDOM.render(
  element,
  document.getElementById('root')
);
```

Because JSX is not JavaScript we will convert it using [Babel](https://babeljs.io/) code compiler. You can get it through the [CDN](https://babeljs.io/docs/en/babel-standalone#script-tags) or using [npm](https://babeljs.io/docs/en/).

## Expressions in JSX

> With JSX you can write expressions using curly braces `{ }`.

Reviewer: ```suggestion
With JSX you can write expressions using curly braces `{ }`. Everything contained inside the curly braces will be evaluated as Javascript code.
```

```js

> const myelement = <h1>I'm { 20 + 1 } years old</h1>;

Reviewer: ```suggestion
const myElement = <h1>I'm { 20 + 1 } years old</h1>;
```

```

## Reserved words

> You must use `className` insted of `class`.

Reviewer: ```suggestion
You must use `className` instead of `class`.  These reserved words are mandatory to avoid clashing with the ones used in HTML.
```

```js

> const myelement = <h1 className='Title'>Hello Push!</h1>;

Reviewer: ```suggestion
const myElement = <h1 className='Title'>Hello Push!</h1>;
```

```

## Properties and methods in JSX are camelCase

```js
const greeting = () => console.log('Hello Push!');

> const myelement = <button onClick={greeting}>Click me!</button>;

Reviewer: ```suggestion
const myElement = <button onClick={greeting}>Click me!</button>;
```

```

## Self-closing tags must end in a slash

```js
const src = '../mypath';
const myImage = <img src={src} alt='my image' />;
```

## Large HTML blocks

Put the HTML inside **parentheses** when working with multiple lines.

```js

> const myelement = (

Reviewer: ```suggestion
const myElement = (
```

  <div>
    <p>Hello</p>
    <p>Push!</p>
  </div>
);
```

## One Top Level Element

> The HTML code must be wrapped only in ONE top level element.

Reviewer: Explain that you `<></>`  can be used to wrap the elements.

> The HTML code must be wrapped only in ONE top level element.

Reviewer: ```suggestion
The HTML code must be wrapped only in ONE top-level element.
```
